src/arbitraries/types.ts

```diff
@@ -8,3 +8,8 @@ export type ArbitrarySize = {
   type: 'exact' | 'estimated'
   credibleInterval?: [number, number]
 }
+
+export type MappedArbitraryConfig<A,B> = {
+  inverseFunction?: (b: B) => A | A[],
+  customCanGenerate?: (pick: FluentPick<B>) => boolean
```

Comment:
I am struggling to understand why you'll need `customCanGenerate`. I don't see it being used here or in #156 and if I understand correctly the meaning of the two configs here, `mappedArbitrary.canGenerate(value)` must be in all cases equal to `inverseFunction(value).some(v => baseArbitrary.canGenerate(v))` - no overrides can exist, except possible for optimized versions (which use cases?).

Reply:
The decision of including both `inverseFunction` and `customCanGenerate` came up from some discussions that me, @arestivo, and @hugoferreira had. I will let them explain their point of view.

Reply:
Copying @arestivo's comment here:

> Sometimes directly writing a _canGenerate_ function can be much easier than going the _inverse_ route.

That makes sense. Do we have any concrete examples of that to justify this more complex API? If not, it might be worth starting with `inverseFunction` and adding it later. Even if we choose to keep it, I would prefer if we disallowed setting _both_ functions (either using the type system or by asserting in runtime) as the issues resulting from inconsistent implementations may be very hard to debug.

Reply:
@ruippeixotog 

```
fc.integer().map(v => utils.sha1(v))
```

Reply:
@ruippeixotog , @arestivo usual gives the example of finding the inverse function of `Math.random()` which is not possible. However, by specifying the `canGenerate` we know that if the value belongs to [0,1[, than it is possible to generate.

Reply:
Sorry, I wasn't clear on the first comment - I was not asking for examples of non-invertible functions, of course :) I was looking for use cases that were meaningful for us as library developers (there are way more efficient ways to generate the space of all SHA1 and generating mapped arbitraries relying on `Math.random()` is a bad use of the library as randomness should be handled by FluentCheck itself).

I was initially assumed that the intention was to expose this to users, as this is the kind of fine-tuning that users don't generally care about (I never saw anything like this in other libraries, at least). As a user, I always saw shrinking as a nice to have but totally optional process for which I wouldn't make my tests more verbose. While getting 1 and "abc" as counterexamples is much nicer than 352853714817 or some long string with unicode chars, bad shrinking never prevented me from finding a root cause.

If you really prefer to expose this (i.e. announce it, not just having it as a hidden features), I'm ok with having both options assuming we assert that users don't provide both.

Reply:
There was extensive discussion about this, and I'm on the "no need for `canGenerate`" team. @arestivo convinced me that coming up with a `canGenerate` can be easier than coming up with an inverse function and I settled for: ok, let's have both. Now @ruippeixotog is convincing me again to move back to the "no need for `canGenerate`" team 😂

I _believe_ that this is not a function for the typical enduser for two reasons:

1. We maintain an history of `original` values with the sole purpose of _not_ needing an inverse when shrinking. We do _need_ a `canGenerate` but those are already provided by the internal arbitraries, including `map` and `filter`. Therefore, for the purpose of shrinking, the _user_ would never need to specify these as long as their arbitraries are a transformation of base arbitraries; do we all agree? 
2. The use case that led to this was the guided search being implemented in #146. We needed a way to find if a certain constant would be _ generateable_ by a certain arbitrary (e.g. `string`'s). But if they were being defined as `map`'s of `integer`'s (e.g. `string`'s), that would not be possible. But defining custom search strategies is not something the typical end-user would be doing...

So, uhhhh, yeah 😅 I would prefer to just have the inverse in map (it's more readable), but I know @arestivo has strong opinions about this 😂

Reply:
> We do *need* a `canGenerate`

We need it only if we consider shrinking to be a requirement rather than a nice to have. It's good if we can provide it, but I'm perfectly ok with not having `canGenerate` for mapped and chained arbitraries built by the user - the added API complexity does not seem worth it.

> Therefore, for the purpose of shrinking, the user would never need to specify these as long as their arbitraries are a transformation of base arbitraries; do we all agree?

Not sure if I agree with this. If we consider shrinking to be a necessity I'm not sure how you'd try to shrink `exists('sha', fc.integer().map(v => utils.sha1(v)).then(sha => sha.endsWith("000"))` (a transformation of base arbitraries) just by looking at the history of values (maybe you can ask the crypto mining community... 🙂 ). If we agree it's a best effort then we really don't need it, but we *never* need it (no "as long as").

Reply:
> I'm perfectly ok with not having canGenerate for mapped and chained arbitraries built by the user

I'm also ok, but the need for `canGenerate` (the method, not the override @RuiGuedes provided) is for shrink to work with arbitraries _built by us_ (e.g. `string`), because they are internally defined as a sequence of `map`'s. We _need it_ to exist, although we don't need it to be _public_. Am I missing something?

> `exists('sha', fc.integer().map(v => utils.sha1(v)).then(sha => sha.endsWith("000"))`

I'm also not sure I agree with you. The example you provide seems to point to a limitation of the search process (finding an hash that ends in `000` is hard), rather than a counter-argument to the internal `canGenerate` definition that is used *after* we've found an example, and that it follows (as you say) a best effort. No?

Reply:
> I'm also ok, but the need for `canGenerate` (the method, not the override @RuiGuedes provided) is for shrink to work with arbitraries _built by us_

Exactly 👍 And hence my question above: do we, library developers, have any use case where a custom `canGenerate` is feasible but `inverseFunction` is not?

> The example you provide seems to point to a limitation of the search process

Yeah, not the best example. I just wanted to illustrate that if we take into consideration everything that end users may try on top of our base arbitraries, you'll find cases where finding an example is easy but there's no obvious or meaningful way to shrink.

Reply:
> Exactly 👍 And hence my question above: do we, library developers, have any use case where a custom `canGenerate` is feasible but `inverseFunction` is not?

My initial answer to your question was _no_. So I'll let @arestivo chime in an present a defense on his perspective 😅

> Yeah, not the best example. I just wanted to illustrate that if we take into consideration everything that end users may try on top of our base arbitraries, you'll find cases where finding an example is easy but there's no obvious or meaningful way to shrink.

Point taken 👍 Once again, the inverse function solved our motivational use case (#146), and it was targeted for users that wanted to extend the library.

Reply:
> The use case that led to this was the guided search being implemented in #146. We needed a way to find if a certain constant would be generateable by a certain arbitrary (e.g. string's). But if they were being defined as map's of integer's (e.g. string's), that wasn't possible. However, defining custom search strategies is not something the typical end-user would be doing...

Just to clear out, indeed the motivation behind the PR was PR #146 as @hugoferreira stated. However, later on, I decided, in PR #156 to enhance the string builder by implementing several other basic arbitraries. When I finished this last PR (#156) I realized that the modifications that the current PR introduce are no longer needed. The reason behind not needing them now is because since `strings` now use ASCII codes it is now possible for a given a string to compute their `original` (array of ASCII codes) value. Therefore the `canGenerate` method can be easily called with `value` and `original`. The way strings were implemented before PR #156 did not allow obtaining the original string value, hence the need for an inverse function.

Considering what I have just said, and considering your last conversation, I'm also on the side of having only the `inverseMap` function.

Reply:
Whatever! I could ask the exact same questions you have about the _inversemap_ method that you ask about the _canGenerate_ method. Let's merge this...

Reply:
@hugoferreira Already managed the conflicts with the `master` branch. However, I think we didn't entirely finish this conversation. How should we proceed xD?

Reply:
Let's settle meet in the middle and at least have the type system disallow specifying both, ok @ruippeixotog and @arestivo? The following code is untested, but I think we can change the constructor of the `MappedArbitrary` to:

```typescript
constructor (..., public readonly shrinkHelper?: { canGenerate: (pick: FluentPick<B>) => boolean} | { inverseMap: (b: B) => A[]}) 
```

Reply:
> > I'm ok with having only `inverseMap`.

Reply:
Ok, I am on it. Tomorrow morning I will finish this PR and believe it can be merged then.

Reply:
@hugoferreira Already addressed the proposed modifications on the last commit including adding some documentation to the `map` function. Feel free to check what I have done, especially the documentation of the `map` function xD. If everything is ok you can approve this and proceed with the merge.